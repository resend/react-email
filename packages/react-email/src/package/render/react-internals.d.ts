declare module 'react-dom/server.browser' {
  export * from 'react-dom/server';
}
