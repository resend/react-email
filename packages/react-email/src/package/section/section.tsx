import * as React from 'react';

export type SectionProps = Readonly<React.ComponentPropsWithoutRef<'table'>>;

export const Section = React.forwardRef<HTMLTableElement, SectionProps>(
  ({ children, style, ...props }, ref) => {
    return (
      <table
        align="center"
        width="100%"
        border={0}
        cellPadding="0"
        cellSpacing="0"
        role="presentation"
        {...props}
        ref={ref}
        style={style}
      >
        <tbody>
          <tr>
            <td>{children}</td>
          </tr>
        </tbody>
      </table>
    );
  },
);

Section.displayName = 'Section';
