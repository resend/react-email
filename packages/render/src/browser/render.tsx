import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type {
  PipeableStream,
  ReactDOMServerReadableStream,
} from 'react-dom/server';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';

const decoder = new TextDecoder('utf-8');

const readStream = async (
  stream: PipeableStream | ReactDOMServerReadableStream,
) => {
  const chunks: Uint8Array[] = [];

  if ('pipeTo' in stream) {
    // means it's a readable stream
    const writableStream = new WritableStream({
      write(chunk: Uint8Array) {
        chunks.push(chunk);
      },
    });
    await stream.pipeTo(writableStream);
  } else {
    throw new Error(
      'For some reason, the Node version of `react-dom/server` has been imported instead of the browser one.',
      {
        cause: {
          stream,
        },
      },
    );
  }

  let length = 0;
  chunks.forEach((item) => {
    length += item.length;
  });
  const mergedChunks = new Uint8Array(length);
  let offset = 0;
  chunks.forEach((item) => {
    mergedChunks.set(item, offset);
    offset += item.length;
  });

  return decoder.decode(mergedChunks);
};

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
