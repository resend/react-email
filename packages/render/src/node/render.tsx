import { convert } from 'html-to-text';
import { Suspense } from 'react';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';
import { pretty } from '../shared/utils/pretty';
import { readStream } from './read-stream';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await import('react-dom/server');

  let html!: string;
  if (Object.hasOwn(reactDOMServer, 'renderToReadableStream')) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
