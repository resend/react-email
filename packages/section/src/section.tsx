import * as React from 'react';

export type SectionProps = Readonly<React.ComponentPropsWithoutRef<'table'>>;

export const Section = React.forwardRef<HTMLTableElement, SectionProps>(
  ({ children, style = {}, ...props }, ref) => {
    // Destructure padding to improve compatibility with Klavyio and Outlook.
    const {
      padding,
      paddingTop,
      paddingRight,
      paddingBottom,
      paddingLeft,
      ...tableStyle
    } = style;

    const tdStyle = {
      padding,
      paddingTop,
      paddingRight,
      paddingBottom,
      paddingLeft,
    };

    return (
      <table
        align="center"
        width="100%"
        border={0}
        cellPadding="0"
        cellSpacing="0"
        role="presentation"
        {...props}
        ref={ref}
        style={tableStyle}
      >
        <tbody>
          <tr>
            <td style={tdStyle}>{children}</td>
          </tr>
        </tbody>
      </table>
    );
  },
);

Section.displayName = 'Section';
