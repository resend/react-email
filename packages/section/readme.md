![React Email Section cover](https://react.email/static/covers/section.png)

<div align="center"><strong>@react-email/section</strong></div>
<div align="center">Display a section that can be formatted using columns.</div>
<br />
<div align="center">
<a href="https://react.email">Website</a> 
<span> · </span>
<a href="https://github.com/resend/react-email">GitHub</a> 
<span> · </span>
<a href="https://react.email/discord">Discord</a>
</div>

## Install

Install component from your command line.

#### With yarn

```sh
yarn add @react-email/section -E
```

#### With npm

```sh
npm install @react-email/section -E
```

## Getting started

Add the component to your email template. Include styles where needed.

```jsx
import { Section } from '@react-email/section';
import { Column } from '@react-email/column';
import { Row } from '@react-email/row';
import { Text } from '@react-email/text';

const Email = () => {
  return (
    {/* A simple `section` */}
    <Section>
      <Text>Hello World</Text>
    </Section>

    {/* Formatted with `rows` and `columns` */}
     <Section>
      <Row>
        <Column>Column 1, Row 1</Column>
        <Column>Column 2, Row 1</Column>
      </Row>
      <Row>
        <Column>Column 1, Row 2</Column>
        <Column>Column 2, Row 2</Column>
      </Row>
    </Section>
  );
};
```

## Border + BorderRadius Compatibility Fix

The Section component now includes automatic handling for `border + borderRadius` combinations to ensure full email client compatibility.

### The Problem

Many email clients have inconsistent support for CSS `border-radius` when used with `border` properties. This can cause rounded corners to not display correctly or borders to appear without the intended rounded corners.

### The Solution

When the Section component detects both `border` and `borderRadius` properties in the style, it automatically wraps the content in a table structure that simulates the border using:

- `backgroundColor` = border color
- `padding` = border width
- `borderRadius` applied to the wrapper table

This approach provides full border-radius support across all email clients.

### Examples

#### Basic border + borderRadius (uses wrapper)
```jsx
<Section
  style={{
    border: '2px solid #e5e7eb',
    borderRadius: '8px',
    padding: '16px',
    backgroundColor: '#f9fafb',
  }}
>
  <p>This will use the border wrapper for compatibility</p>
</Section>
```

#### Only border (no wrapper needed)
```jsx
<Section
  style={{
    border: '1px solid #d1d5db',
    padding: '16px',
  }}
>
  <p>This renders normally without wrapper</p>
</Section>
```

#### Only borderRadius (no wrapper needed)
```jsx
<Section
  style={{
    borderRadius: '8px',
    padding: '16px',
    backgroundColor: '#fef3c7',
  }}
>
  <p>This renders normally without wrapper</p>
</Section>
```

#### Individual border properties
```jsx
<Section
  style={{
    borderWidth: '1px',
    borderStyle: 'solid',
    borderColor: '#3b82f6',
    borderRadius: '12px',
    padding: '20px',
  }}
>
  <p>This will use the border wrapper for compatibility</p>
</Section>
```

### How It Works

1. **Detection**: The component checks if both border and borderRadius properties are present
2. **Wrapper Application**: If detected, content is wrapped in a table with:
   - Background color matching the border color
   - Padding equal to the border width
   - Border radius applied to the wrapper
3. **Style Preservation**: Non-border styles are preserved on the inner element
4. **Fallback**: If no border + borderRadius combination is detected, normal rendering occurs

### Supported Border Properties

The fix detects and handles:
- `border` (shorthand)
- `borderTop`, `borderRight`, `borderBottom`, `borderLeft`
- `borderWidth`, `borderStyle`, `borderColor`
- `borderRadius` (shorthand)
- `borderTopLeftRadius`, `borderTopRightRadius`, `borderBottomLeftRadius`, `borderBottomRightRadius`

### Email Client Compatibility

This approach ensures consistent border-radius rendering across:
- Gmail (all platforms)
- Outlook (all versions)
- Apple Mail
- Yahoo Mail
- Thunderbird
- And other major email clients

## Support

This component was tested using the most popular email clients.

| <img src="https://react.email/static/icons/gmail.svg" width="48px" height="48px" alt="Gmail logo"> | <img src="https://react.email/static/icons/apple-mail.svg" width="48px" height="48px" alt="Apple Mail"> | <img src="https://react.email/static/icons/outlook.svg" width="48px" height="48px" alt="Outlook logo"> | <img src="https://react.email/static/icons/yahoo-mail.svg" width="48px" height="48px" alt="Yahoo! Mail logo"> | <img src="https://react.email/static/icons/hey.svg" width="48px" height="48px" alt="HEY logo"> | <img src="https://react.email/static/icons/superhuman.svg" width="48px" height="48px" alt="Superhuman logo"> |
| -------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------ |
| Gmail ✔                                                                                           | Apple Mail ✔                                                                                           | Outlook ✔                                                                                             | Yahoo! Mail ✔                                                                                                | HEY ✔                                                                                         | Superhuman ✔                                                                                                |

## License

MIT License
