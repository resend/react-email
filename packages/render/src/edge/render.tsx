import { Suspense } from 'react';
import { pretty } from '../node';
import type { Options } from '../shared/options';
import { readStream } from '../shared/read-stream.browser';
import { toPlainText } from '../shared/utils/to-plain-text';
import { importReactDOM } from './import-react-dom';

export const render = async (
  element: React.ReactElement,
  options?: Options,
) => {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const reactDOMServer = await importReactDOM().then(
    // This is beacuse react-dom/server is CJS
    (m) => m.default,
  );
  console.log(reactDOMServer);

  const html = await new Promise<string>((resolve, reject) => {
    reactDOMServer
      .renderToReadableStream(suspendedElement, {
        onError(error: unknown) {
          reject(error);
        },
        progressiveChunkSize: Number.POSITIVE_INFINITY,
      })
      .then(readStream)
      .then(resolve)
      .catch(reject);
  });

  if (options?.plainText) {
    return toPlainText(html, options.htmlToTextOptions);
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
