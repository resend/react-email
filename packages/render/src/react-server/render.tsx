import { convert } from 'html-to-text';
import { Suspense } from 'react';
import { experimental_renderToHTML as renderToHTML } from 'react-markup';
import { pretty } from '../node';
import type { Options } from '../shared/options';
import { plainTextSelectors } from '../shared/plain-text-selectors';

export const render = async (node: React.ReactNode, options?: Options) => {
  const suspendedElement = <Suspense>{node}</Suspense>;

  const html = await renderToHTML(suspendedElement);

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, '')}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
};
