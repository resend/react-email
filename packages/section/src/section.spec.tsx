import { render } from '@react-email/render';
import { Section } from './section.js';

describe('Section component', () => {
  it('should render correctly', async () => {
    const actualOutput = await render(
      <Section>
        <div>Test content</div>
      </Section>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should render with style', async () => {
    const actualOutput = await render(
      <Section style={{ backgroundColor: 'red' }}>
        <div>Test content</div>
      </Section>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should render with props', async () => {
    const actualOutput = await render(
      <Section align="center" width="100%">
        <div>Test content</div>
      </Section>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should use BorderWrapper when both border and borderRadius are present', async () => {
    const actualOutput = await render(
      <Section style={{ 
        border: '1px solid black', 
        borderRadius: '8px',
        color: 'red' 
      }}>
        <div>Test content</div>
      </Section>,
    );
    
    // Should contain the wrapper table with background-color and padding
    expect(actualOutput).toContain('background-color:black');
    expect(actualOutput).toContain('padding:1');
    expect(actualOutput).toContain('border-radius:8px');
    expect(actualOutput).toContain('<div>Test content</div>');
    expect(actualOutput).not.toContain('border:1px solid black');
  });

  it('should not use BorderWrapper when only border is present', async () => {
    const actualOutput = await render(
      <Section style={{ 
        border: '1px solid black',
        color: 'red' 
      }}>
        <div>Test content</div>
      </Section>,
    );
    
    // Should render normally without wrapper
    expect(actualOutput).toContain('border:1px solid black');
    expect(actualOutput).not.toContain('background-color:black');
    expect(actualOutput).not.toContain('padding:1');
  });

  it('should not use BorderWrapper when only borderRadius is present', async () => {
    const actualOutput = await render(
      <Section style={{ 
        borderRadius: '8px',
        color: 'red' 
      }}>
        <div>Test content</div>
      </Section>,
    );
    
    // Should render normally without wrapper
    expect(actualOutput).toContain('border-radius:8px');
    expect(actualOutput).not.toContain('background-color:');
    expect(actualOutput).not.toContain('padding:');
  });

  it('should handle individual border properties with borderRadius', async () => {
    const actualOutput = await render(
      <Section style={{ 
        borderWidth: '2px',
        borderStyle: 'solid',
        borderColor: 'red',
        borderRadius: '4px',
        color: 'blue' 
      }}>
        <div>Test content</div>
      </Section>,
    );
    
    expect(actualOutput).toContain('background-color:red');
    expect(actualOutput).toContain('padding:2');
    expect(actualOutput).toContain('border-radius:4px');
    expect(actualOutput).toContain('color:blue');
  });
});
