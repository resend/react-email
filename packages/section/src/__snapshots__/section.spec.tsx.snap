// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Section> component > renders correctly 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation"><tbody><tr><td>Lorem ipsum</td></tr></tbody></table><!--/$-->"`;

exports[`Section component > should render correctly 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation"><tbody><tr><td><div>Test content</div></td></tr></tbody></table><!--/$-->"`;

exports[`Section component > should render with props 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation"><tbody><tr><td><div>Test content</div></td></tr></tbody></table><!--/$-->"`;

exports[`Section component > should render with style 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="background-color:red"><tbody><tr><td><div>Test content</div></td></tr></tbody></table><!--/$-->"`;
